cmgd/cmgd_db.c

```diff
@@ -44,62 +46,89 @@
 
 typedef struct cmgd_db_ctxt_ {
         cmgd_database_id_t db_id;
+		pthread_rwlock_t rw_lock;
 
-        struct nb_node *root;
+        struct nb_config *root;
 } cmgd_db_ctxt_t;
 
 int cmgd_db_init(struct cmgd_master *cm)
 {
-	// TODO: Initialize cm->running_db, cm->candidate_db, cm->oper_db.
+	cmgd_db_ctxt_t running, candidate, oper;
+
+	// db_id to be added
+	running.root = nb_config_new(NULL);
+	candidate.root = nb_config_new(NULL);
+	oper.root = nb_config_new(NULL);
+
+	cm->running_db = (cmgd_db_hndl_t)&running;
+	cm->cnadidate_db = (cmgd_db_hndl_t)&candidate;
+	cm->oper_db = (cmgd_db_hndl_t)&oper;
+
 	return 0;
 }
 
-void cmgd_db_read_lock(cmgd_db_hndl_t db_hndl)
+int cmgd_db_read_lock(cmgd_db_hndl_t db_hndl)
 {
 	cmgd_db_ctxt_t *db_ctxt;
+	int lock_status;
 
 	db_ctxt = (cmgd_db_ctxt_t *)db_hndl;
 	if (!db_ctxt)
-		return;
+		return -1;
+
+	lock_status = pthread_rwlock_tryrdlock(&db_ctxt->rw_lock);
+	return lock_status;
 
-	/* pthread_rw_rdlock(db_ctxt->rw_lock) */
 }
 
-void cmgd_db_write_lock(cmgd_db_hndl_t db_hndl)
+int cmgd_db_write_lock(cmgd_db_hndl_t db_hndl)
 {
 	cmgd_db_ctxt_t *db_ctxt;
+	int lock_status;
 
 	db_ctxt = (cmgd_db_ctxt_t *)db_hndl;
 	if (!db_ctxt)
-		return;
+		return -1;
 
-	/* pthread_rw_wrlock(db_ctxt->rw_lock) */
+	lock_status = pthread_rwlock_trywrlock(&db_ctxt->rw_lock);
+	return lock_status;
 }
 
-void cmgd_db_unlock(cmgd_db_hndl_t db_hndl)
+int cmgd_db_unlock(cmgd_db_hndl_t db_hndl)
 {
 	cmgd_db_ctxt_t *db_ctxt;
+	int lock_status;
 
 	db_ctxt = (cmgd_db_ctxt_t *)db_hndl;
 	if (!db_ctxt)
-		return;
+		return -1;
 
-	/* pthread_rw_unlock(db_ctxt->rw_lock) */
+	lock_status =  pthread_rwlock_unlock(&db_ctxt->rw_lock);
+	return lock_status;
 }
 
 int cmgd_db_lookup_data_nodes(
         cmgd_db_hndl_t db_hndl, const char *xpath,
-        struct lyd_node *dnode[], struct nb_node *nb_node[], 
-        int *num_nodes)
+        struct lyd_node *dnodes[], int *num_nodes)
 {
 	cmgd_db_ctxt_t *db_ctxt;
+	struct ly_set *set = NULL;
+	uint32_t i;
 
 	db_ctxt = (cmgd_db_ctxt_t *)db_hndl;
 	if (!db_ctxt)
 		return -1;
 
-	dnode[0] = db_ctxt->root;
-	*num_nodes = 1;
+	if (xpath[0] == '.' && xpath[1] == '/')
+		xpath += 2;
+
+	lyd_find_xpath(db_ctxt->root->dnode, xpath, &set);
```

Comment:
Can you not have used LYD_TREE_DFS_BEGIN() and LYD_TREE_DFS_END() macros to iterate and collect the dnodes[] and num_nodes?